File: This Is Coding Test with Python/C++/ThisIsCodingTest-Basic/Greedy/[3-4] 1이 될 때까지.cpp
#include <iostream>
#include <vector>
#include "CodingTester.h"

using namespace std;

struct Param {
	int n{};
	int k{};

	friend istream& operator>>( istream& is, Param& self )
	{
		is >> self.n >> self.k;

		return is;
	}
};

struct TestSet {
	int num{};
	Param param{};
	int result{};

	TestSet() = default;
	TestSet( Param p, int r ) {
		TestSet();
		param = p;
		result = r;
	}
	friend istream& operator>>( istream& is, TestSet& t )
	{
		is >> t.param;
		is >> t.result;

		return is;
	}
};

template <>
struct std::formatter<TestSet> {
	constexpr auto parse( format_parse_context& ctx ) { return ctx.begin(); }

	template <typename FormatContext>
	auto format( const TestSet& ts, FormatContext& ctx ) {

		string strnum = "[" + to_string( ts.num ) + "]";
		auto out = format_to( ctx.out(), " {:4} | ", strnum );

		out = format_to( out, "n:{}, k:{}\n", ts.param.n, ts.param.k );
		out = format_to( out, "{:5} | ", "" );
		out = format_to( out, "Result: {}", ts.result );

		return out;
	}
};

int MySolution( Param param );
int BookSolution( Param param );

int main()
{
	auto test_sets{ ReadTestFile<TestSet>( "../../../TestSets/3-4.txt" ) };

	for ( int i{}; const auto & test_set : test_sets ) {
		OutputTestSolution<Param, int, TestSet>( MySolution, ++i, test_set.param, test_set.result );
		cout << endl;
	}

	//for ( int i{}; const auto & test_set : test_sets ) {
	//	OutputTestSolution<Param, int, TestSet>( BookSolution, ++i, test_set.param, test_set.result );
	//	cout << endl;
	//}
}

int MySolution( Param param )
{
	int dividend = param.n;
	int divisor = param.k;
	int remainder{}, amount{};
	
	do {
		remainder = dividend % divisor;
		dividend /= divisor;
		amount += remainder + (dividend < 1 ? -1 : 1);
	} while ( dividend > 1 );

	return amount;
}

int BookSolution( Param param )
{
	return 0;
}

File: This Is Coding Test with Python/Python/ThisIsCodingTest/Greedy/[3-4] 1이 될 때까지.py
class Param:
    def __init__(self, values: list):
        self.n, self.k, = values


def MySolution(param: Param):
    dividend = param.n
    divisor = param.k
    amount, remainder = 0, 0

    while dividend > 1:
        remainder = dividend % divisor
        dividend //= divisor
        amount += remainder + (1 if dividend >= 1 else -1)

        # if dividend <= 1:
        #     break

    return amount


def BookSolution(param):
    result = 0
    n, k = param.n, param.k

    while True:
        target = (n // k) * k
        result += (n - target)
        n = target

        if n < k:
            break

        result += 1
        n //= k

    return result + (n - 1)


if __name__ == "__main__":
    import CodingTester

    class TestSet:
        def __init__(self, num, p: list, r):
            self.num = num
            self.param: Param
            if p.__class__ is list:
                self.param = Param(p)
            elif p.__class__ is Param:
                self.param = p

            self.result = r

        def __format__(self, format_spec):
            num_str = "[" + str(self.num) + "]"
            a = f" {num_str:4} | n: {self.param.n}, k: {self.param.k}\n"
            b = f"      | result: {self.result}"
            return a + b

    file = CodingTester.ReadTestFile("../../../TestSets/3-4.txt")
    test_sets = list()

    i = 0
    while i < len(file):
        test_sets.append(TestSet(0, file[i:i+2], file[i+2]))
        i += 3

    print("My Solution =======>")
    for i, test_set in enumerate(test_sets):
        CodingTester.OutputTestSolutionExt(MySolution, i + 1, test_set.param, test_set.result, TestSet)
        print()

    print("Book's Solution =======>")
    for i, test_set in enumerate(test_sets):
        CodingTester.OutputTestSolutionExt(BookSolution, i + 1, test_set.param, test_set.result, TestSet)
        print()
